Sources/NIOHTTP1/HTTPUpgradeHandler.swift

```diff
@@ -108,74 +108,99 @@ public class HTTPServerUpgradeHandler: ChannelInboundHandler, RemovableChannelHa
+        case .awaitingUpgrader:
+            if case .end = requestPart {
+                // This is the end of the first request. Swallow it, we're buffering the rest.
                 self.seenFirstRequest = true
-
-                // The request is complete now trigger the upgrade.
-                upgrade()
-            }
-        } else {
-            // We should decide if we're going to upgrade based on the first request header: if we aren't upgrading,
-            // by the time the body comes in we should be out of the pipeline. That means that if we don't think we're
-            // upgrading, the only thing we should see is a request head. Anything else in an error.
-            guard case .head(let request) = requestPart else {
-                context.fireErrorCaught(HTTPServerUpgradeErrors.invalidHTTPOrdering)
-                notUpgrading(context: context, data: data)
-                return
             }
-            
-            // Ok, we have a HTTP request. Check if it's an upgrade. If it's not, we want to pass it on and remove ourselves
-            // from the channel pipeline.
-            let requestedProtocols = request.headers[canonicalForm: "upgrade"]
-            guard requestedProtocols.count > 0 else {
-                notUpgrading(context: context, data: data)
-                return
+        case .upgraderReady(let upgrade):
+            if case .end = requestPart {
+                // This is the end of the first request, and we can upgrade. Time to kick it off.
+                self.seenFirstRequest = true
+                upgrade()
             }
-            
-            // Cool, this is an upgrade! Let's go.
-            if let upgrade = handleUpgrade(context: context, request: request, requestedProtocols: requestedProtocols) {
-                self.upgrade = upgrade
+        case .upgradeFailed:
+            // We were re-entrantly called while delivering the request head. We can just pass this through.
+            context.fireChannelRead(data)
+        case .upgradeComplete:
+            preconditionFailure("Upgrade has completed but we have not seen a whole request and still got re-entrantly called.")
+        case .upgrading:
+            preconditionFailure("We think we saw .end before and began upgrading, but somehow we have not set seenFirstRequest")
+        }
+    }
+
+    private func firstRequestHeadReceived(context: ChannelHandlerContext, requestPart: HTTPServerRequestPart) {
+        // We should decide if we're going to upgrade based on the first request header: if we aren't upgrading,
+        // by the time the body comes in we should be out of the pipeline. That means that if we don't think we're
+        // upgrading, the only thing we should see is a request head. Anything else in an error.
+        guard case .head(let request) = requestPart else {
+            context.fireErrorCaught(HTTPServerUpgradeErrors.invalidHTTPOrdering)
```

Comment:
is this not a precondition?

Reply:
I'm open to it being one. This isn't a change in behaviour from the old model: we also treated it as an error there. I think I was reluctant to precondition on *inbound* data, but I'm open to deciding that there are enough handlers between here and the network that it's safe to do so.

Reply:
Ok, let’s leave fof now